### PR TITLE
COCICCS-4439 ,  COCICCS-4455 , COCICCS-4756  - Combined PR 

### PR DESCRIPTION
**Pull request by NNazeer1**
Onboard Azure API, List AAD Service Principals , Read AAD Service Principal Secret

### CODE REVIEW

**Review by alexrajan3 — DISMISSED** [review]

File: tvaultapi/src/main/resources/templates/AzureEmailTemplate.html [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        <li>
+            <div>Navigate to the "MANAGE" tab</div>
+            <img class ="image" src="iammanagetab.png" th:src="|cid:${iammanagetab}|"/>
+        </li>
+        <li>
+            <div>Select the "View/Edit" option corresponding to the Azure service account you want to activate.</div>
+            <img class ="image" src="iamviewlink.png" th:src="|cid:${iamviewlink}|"/>
+        </li>
+        <li>
+            <div>Click the "Activate IAM Service Account" link at the bottom right.</div>
```
Comment (sarathmohan4) [inline tvaultapi/src/main/resources/templates/AzureEmailTemplate.html:32]: IAM

File: tvaultui/src/app/Features/azure-service-prinicipals/azure-service-prinicipals.controller.js [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+                });
+        }
+
+        $scope.viewSecret = function (azuresvcaccname, folderName) {
+            var path = azuresvcaccname + '/' + folderName;
+            $scope.isLoadingData = true;
+            $scope.write = false;
+            $scope.azuresvcaccSecretData = {};
+            var updatedUrlOfEndPoint = RestEndpoints.baseURL + "/v2/azureserviceaccounts/secrets/" + path;
+            AdminSafesManagement.getSecretForIamSvcacc(null, updatedUrlOfEndPoint).then(function (response) {
```
Comment (sarathmohan4) [inline tvaultui/src/app/Features/azure-service-prinicipals/azure-service-prinicipals.controller.js:146]: is this getSecretForIamSvcacc() ? correct?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java [hunk 3]
```diff
@@ -0,0 +1,1357 @@
...
+	/**
+	 * Method to populate AzureServiceAccountMetadataDetails object
+	 *
+	 * @param azureServiceAccount
+	 * @return
+	 */
+	private AzureServiceAccountMetadataDetails constructAzureSvcAccMetaData(AzureServiceAccount azureServiceAccount) {
+
+		AzureServiceAccountMetadataDetails azureServiceAccountMetadataDetails = new AzureServiceAccountMetadataDetails();
+		List<AzureSecretsMetadata> iamSecretsMetadatas = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (NNazeer1) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java:301]: done

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java [hunk 4]
```diff
@@ -0,0 +1,1357 @@
...
+					.append(AzureServiceAccountConstants.AZURE_SVCACC_POLICY_PREFIX).append(azureSvcAccName).toString();
+			accessPolicy.setAccessid(accessId);
+			HashMap<String, String> accessMap = new HashMap<>();
+			String iamCredsPath=new StringBuffer().append(AzureServiceAccountConstants.AZURE_SVCC_ACC_PATH).append(azureSvcAccName).append("/*").toString();
+			accessMap.put(iamCredsPath, TVaultConstants.getSvcAccPolicies().get(policyPrefix));
+			// Attaching write permissions for owner
+			if (TVaultConstants.getSvcAccPolicies().get(policyPrefix).equals(TVaultConstants.SUDO_POLICY)) {
+				accessMap.put(AzureServiceAccountConstants.AZURE_SVCC_ACC_PATH + azureSvcAccName + "/*",
+						TVaultConstants.WRITE_POLICY);
+				accessMap.put(AzureServiceAccountConstants.AZURE_SVCC_ACC_PATH + azureSvcAccName,
```
[the remark this answers is not in the document]
Comment (NNazeer1) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java:417]: done

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java [hunk 5]
```diff
@@ -0,0 +1,1357 @@
...
+						TVaultConstants.WRITE_POLICY);
+			}
+			if (TVaultConstants.getSvcAccPolicies().get(policyPrefix).equals(TVaultConstants.WRITE_POLICY)) {
+				accessMap.put(AzureServiceAccountConstants.AZURE_SVCC_ACC_PATH + azureSvcAccName + "/*",
+						TVaultConstants.WRITE_POLICY);
+				accessMap.put(AzureServiceAccountConstants.AZURE_SVCC_ACC_PATH + azureSvcAccName,
```
[the remark this answers is not in the document]
Comment (NNazeer1) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java:423]: done

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java [hunk 6]
```diff
@@ -0,0 +1,1357 @@
...
+			userResponse.setResponse(oidcEntityResponse.getPolicies().toString());
+			userResponse.setHttpstatus(responseEntity.getStatusCode());
+		}
+
+		return addPolicyToAzureSvcAcc(token, userDetails, azureServiceAccountUser, oidcEntityResponse, azureSvcaccName,
+				userResponse);
+	}
+	
+	/**
+	 * Method to create policies for add user to IAM service account and call the update process.
```
[the remark this answers is not in the document]
Comment (NNazeer1) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java:873]: done

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java [hunk 7]
```diff
@@ -0,0 +1,1357 @@
...
+		}
+		if (path != null && path.endsWith("/")) {
+			path = path.substring(0, path.length() - 1);
+		}
+		String azureMetaDataPath = "metadata/" + path;
+		return reqProcessor.process("/sdb", "{\"path\":\"" + azureMetaDataPath + "\"}", token);
+	}
+	
+	/**
+	 * Method to verify the user for add user to IAM service account.
```
[the remark this answers is not in the document]
Comment (NNazeer1) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java:826]: done

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java [hunk 8]
```diff
@@ -0,0 +1,1357 @@
...
+			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
+					.body("{\"errors\":[\"Failed to add user to the Azure Service Account\"]}");
+		}
+	}
+	
+	/**
+	 * Method to update metadata for add user to IAM service account.
+	 * @param token
+	 * @param userDetails
+	 * @param iamServiceAccountUser
```
[the remark this answers is not in the document]
Comment (NNazeer1) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java:1009]: done

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java [hunk 9]
```diff
@@ -0,0 +1,1357 @@
...
+					"{\"messages\":[\"Failed to add user to the Service Account. Metadata update failed\"]}");
+		} else {
+			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
+					.body("{\"messages\":[\"Failed to revert user association on Azure Service Account\"]}");
+		}
+	}
+	
+	/**
+	 * Method to check if the owner permission is getting changed.
+	 * @param iamServiceAccountUser
```
[the remark this answers is not in the document]
Comment (NNazeer1) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/AzureServicePrinicipalAccountsService.java:1101]: done

File: tvaultapi/src/main/resources/templates/AzureEmailTemplate.html (hunk 1 above)
Comment (NNazeer1) [reply tvaultapi/src/main/resources/templates/AzureEmailTemplate.html:32]: done

File: tvaultui/src/app/Features/azure-service-prinicipals/azure-service-prinicipals.controller.js (hunk 2 above)
Comment (NNazeer1) [reply tvaultui/src/app/Features/azure-service-prinicipals/azure-service-prinicipals.controller.js:146]: done

**Review by sarathmohan4 — APPROVED** [review]

**Pull request merged**